### ISSUE TITLE
OpenCL kernel argument type mismatch

### ISSUE THREAD

**Issue by thtrummer** [openmm/openmm#2699]
[This line](https://github.com/openmm/openmm/blob/fdfdb0f0d680ff7e211b3e93e399712a9d11561c/platforms/common/src/IntegrationUtilities.cpp#L552) adds an integer argument (size==4) for a pointer argument (`GLOBAL real4* RESTRICT posqCorrection`). This fails at runtime when building for 64 bit.
```
Function: clSetKernelArg
Error:    CL_INVALID_ARG_SIZE
arg_size is 4, but argument should be 8 bytes
```

**Comment by peastman** [comment]
What platform are you running on?  The only way I see that happening is if your host code were compiled in 32 bit mode but device code were being compiled in 64 bit mode.

**Comment by thtrummer** [comment]
The platform is 64 bit Windows.

`NULL` in C++ is an [integer literal](https://en.cppreference.com/w/cpp/types/NULL) not a pointer type so ` addArg` is instantiated as 

```
void addArg(const int& value) {
    addPrimitiveArg(&value, sizeof(int));
}
```

**Comment by peastman** [comment]
According to that page, since C++11 it's been a `nullptr_t` (which is a pointer type), not an int.  What compiler are you using?

**Comment by thtrummer** [comment]
Emphasis mine:
> an integer literal with value zero, **or** a prvalue of type std::nullptr_t

Also [`std::nullptr_t`](https://en.cppreference.com/w/cpp/types/nullptr_t) is not a pointer type:
> std::nullptr_t is the type of the null pointer literal, nullptr. It is a distinct type that is not itself a pointer type or a pointer to member type.

**Comment by peastman** [comment]
What compiler are you using?  I'd like to try to reproduce this.

**Comment by thtrummer** [comment]
Visual Studio 2019
```
λ cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28805 for x64
```
Here's the definition of NULL (from `vcruntime.h`):
```
#ifndef NULL
    #ifdef __cplusplus
        #define NULL 0
    #else
        #define NULL ((void *)0)
    #endif
#endif
```

**Comment by peastman** [comment]
How irritating!  It looks like C++11 introduced a new keyword `nullptr` as a replacement for `NULL`.  If you use that does it work?

**Comment by thtrummer** [comment]
It does work because it must have the same size as `void*` (assuming all pointer types have the same size, which is not guaranteed but a reasonable assumption). From an interface perspective this still seems brittle.

> **6.9.1  Fundamental types**
[...]
A value of type `std::nullptr_t` is a null pointer constant (7.11). Such values participate in the pointer and the pointer to member conversions (7.11, 7.12). `sizeof(std::nullptr_t)` shall be equal to `sizeof(void*)`.

(If you get ever bitten by the bug to replace all instances of `NULL`with `nullptr`, there are tools like [`clang-tidy`](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-nullptr.html) that can do that automatically.)

**Comment by peastman** [comment]
> From an interface perspective this still seems brittle

In what way?

**Comment by thtrummer** [comment]
> In what way?

People in the future have to know/remember that they have to write `addArg(nullptr)` instead of `addArg(NULL)` without the latter being a compile error.

**Comment by peastman** [comment]
The value you're passing is the value that should be passed to the kernel.  When we write `nullptr` there, we're literally saying that the kernel argument is a pointer, and the value that should be passed for that argument is a null pointer.

**Comment by peastman** [comment]
Fixed by #2704.
